### PR TITLE
BattleShip-3 - Include Facebook API it to project

### PR DESCRIPTION
**Pull request by kolokolochek**
* added figaro gem
* rename all authorization files to session files
* ability to sign in via facebook
* created user model
* added indexing for user tible
* added model validations
* added messages to yml file
* added test coverage

### CODE REVIEW

File: spec/controllers/application_controller_spec.rb [hunk 1]
```diff
@@ -14,8 +20,36 @@
...
+    it 'can use FactoryBot' do
+      expect(current_user).to be_present
+    end
+  end
+
+  describe '.authenticate' do
+    before { get :index }
+
+    it 'should be called before render page' do
+      allow(described_class.new).to receive(:authenticate)
```
Comment (kolokolochek) [inline spec/controllers/application_controller_spec.rb]: Does it make sense?

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -2,4 +2,18 @@
 
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+
+  before_action :authenticate
+
+  helper_method :current_user
```
Comment (IlyaUmanets) [inline app/controllers/application_controller.rb]: Write `helper_method` below method definition

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    redirect_to log_in_path, notice: I18n.t('messages.log_out')
+  end
+
+  def failure
+    redirect_to log_in_path
+  end
+
+  private
+
+  def auth_params(params = request.env['omniauth.auth'])
```
Comment (IlyaUmanets) [inline app/controllers/sessions_controller.rb]: No need to set `params` param here
Define this var in a method body

File: app/views/layouts/application.slim [hunk 4]
```diff
@@ -1,11 +1,13 @@
...
-    meta content=('text/html; charset=UTF-8') http-equiv='Content-Type' /
+    meta content=('text/html; charset=UTF-8') http-equiv='Content-Type'
     title BattleShip
     = csrf_meta_tags
     = stylesheet_link_tag    'application', media: 'all', 'data-turbolinks-track': 'reload'
     = javascript_include_tag 'application', 'data-turbolinks-track': 'reload'
   body
     .page-wrapper
+      = notice
+      = link_to 'LogOut', sign_out_path, method: :delete
```
Comment (IlyaUmanets) [inline app/views/layouts/application.slim]: Use I18n please

File: config/application.example.yml [hunk 5]
```diff
@@ -0,0 +1,2 @@
+FACEBOOK_KEY: facebook-id
```
Comment (IlyaUmanets) [inline config/application.example.yml]: No need to set a value here. Leave this values empty

File: config/locales/en.yml [hunk 6]
```diff
@@ -1,2 +1,7 @@
 en:
-  hello: "Hello world"
+  url:
+    image:
+      facebook: 'http://graph.facebook.com/v2.6/%{uid}/picture'
+  messages:
+    log_in: 'You have successfully logged in.'
+    log_out: 'You have been logged off. See You again soon.'
```
Comment (IlyaUmanets) [inline config/locales/en.yml]: Please add an empty line here

File: config/routes.rb [hunk 7]
```diff
@@ -1,5 +1,9 @@
 # frozen_string_literal: true
 
 Rails.application.routes.draw do
-  get '/log_in' => 'authorizations#log_in'
+  root 'application#index'
+
+  get    '/log_in',                  to: 'sessions#log_in'
```
Comment (IlyaUmanets) [inline config/routes.rb]: `get 'log_in'`

File: config/routes.rb [hunk 8]
```diff
@@ -1,5 +1,9 @@
...
+  get    '/auth/:provider/callback', to: 'sessions#create'
+  delete '/sign_out',                to: 'sessions#destroy'
```
Comment (IlyaUmanets) [inline config/routes.rb]: same

File: spec/controllers/application_controller_spec.rb [hunk 9]
```diff
@@ -3,7 +3,13 @@
 require 'rails_helper'
 
 RSpec.describe ApplicationController, type: :controller do
-  let(:fake_user) { build(:fake_user) }
+  let(:current_user) { create(:user) }
+
+  describe 'routing' do
```
Comment (IlyaUmanets) [inline spec/controllers/application_controller_spec.rb]: This spec does not make sense. You don't need to test Rails or other gems

File: spec/controllers/application_controller_spec.rb [hunk 10]
```diff
@@ -14,8 +20,36 @@
       expect { Faker::Name.name }.not_to raise_error
     end
 
-    skip 'can use FactoryBot' do
-      expect { fake_user }.to be_present
+    it 'can use FactoryBot' do
```
Comment (IlyaUmanets) [inline spec/controllers/application_controller_spec.rb]: This spec does not make sense to. You are trying to test that Factory Bot works well.

File: spec/controllers/application_controller_spec.rb [hunk 11]
```diff
@@ -14,8 +20,36 @@
...
+      expect(current_user).to be_present
+    end
+  end
+
+  describe '.authenticate' do
```
Comment (IlyaUmanets) [inline spec/controllers/application_controller_spec.rb]:
```ruby
describe '#authenticate' do
  it 'should be called before render page' do
    expect_any_instance_of(described_class).to receive(:authenticate)

    get :index
  end

  it 'should redirect not authorized user to log in page' do
    get :index

    expect(response).to redirect_to(log_in_path)
  end

  it 'should not redirect authorized user to log in page', authenticate: true do
    get :index

    expect(response).to render_template(:index)
  end
end

describe '#current_user' do
  let(:instance) { described_class.new }

  skip 'should return nil if empty session' do
    expect(instance.current_user).to be_nil
  end

  skip 'should return user by id from session', authenticate: true do
    expect(instance.current_user).to eq(current_user)
  end
end
```

File: spec/controllers/sessions_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe SessionsController, type: :controller do
+  let(:current_user) { create(:user) }
+
+  describe 'routing' do
+    it 'should route requests to correct actions' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: It does not make sense

File: spec/controllers/sessions_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+      expect(get: '/log_in').to route_to(controller: 'sessions', action: 'log_in')
+      expect(get: '/auth/facebook/callback').to route_to(controller: 'sessions', action: 'create', provider: 'facebook')
+      expect(delete: '/sign_out').to route_to(controller: 'sessions', action: 'destroy')
+    end
+  end
+
+  context 'not Authorized' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: when use is not authorized

File: spec/controllers/sessions_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+    describe '.log_in' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: `#log_in`

File: spec/controllers/sessions_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+      it 'can see log in page' do
+        get :log_in
+
+        expect(response).to render_template(:log_in)
+      end
+    end
+
+    describe '.create' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: `#create`

File: spec/models/user_spec.rb [hunk 16]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  let(:user)   { create(:user) }
+  let(:fields) do
+    {
+      name: Faker::Name.name,
```
Comment (IlyaUmanets) [inline spec/models/user_spec.rb]: use `attributes_for(:user)`

File: spec/models/user_spec.rb [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+  describe 'attributes' do
+    it { is_expected.to respond_to :name }
+    it { is_expected.to respond_to :uid }
+    it { is_expected.to respond_to :provider }
+    it { is_expected.to respond_to :image }
+  end
+
+  describe 'field validations' do
+    it { is_expected.to validate_presence_of(:name).with_message("can't be blank") }
```
Comment (IlyaUmanets) [inline spec/models/user_spec.rb]: Use this [gem](https://github.com/thoughtbot/shoulda-matchers)

Comment (kolokolochek) [reply spec/models/user_spec.rb]: Already used [here](https://github.com/kolokolochek/BattleShip/pull/7/files#diff-8b7db4d5cc4b8f6dc8feb7030baa2478R26).

File: config/application.example.yml [hunk 18]
```diff
@@ -0,0 +1,2 @@
+FACEBOOK_KEY:
```
Comment (IlyaUmanets) [inline config/application.example.yml]: `FACEBOOK_KEY: ''`

File: spec/controllers/application_controller_spec.rb [hunk 19]
```diff
@@ -3,19 +3,37 @@
...
+      get :index
+
+      expect(response).to render_template(:index)
+    end
+  end
+
+  describe '#current_user' do
+    let(:instance) { described_class.new }
+
+    skip 'should return nil if empty session' do
```
Comment (IlyaUmanets) [inline spec/controllers/application_controller_spec.rb]: We don't need to keep `skip` specs in a `master` branch

File: spec/controllers/sessions_controller_spec.rb [hunk 20]
```diff
@@ -0,0 +1,80 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe SessionsController, type: :controller do
+  let(:current_user) { create(:user) }
+
+  context 'when user is not authorized' do
+    describe '#log_in' do
+      it 'can see log in page' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: `log_in`

File: spec/controllers/sessions_controller_spec.rb [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+        get :log_in
+
+        expect(response).to render_template(:log_in)
+      end
+    end
+
+    describe '#create' do
+      skip 'should be redirected to log in page' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: same about `skip` and `log_in`

File: spec/controllers/sessions_controller_spec.rb [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+        expect(flash.notice).to eq('You have been logged off. See You again soon.')
+      end
+    end
+  end
+
+  context 'when user is authorized', authenticate: true do
+    describe '#log_in' do
+      before { get :log_in }
+
+      it 'should be redirected to root page' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: `it 'redirects to root page'`

File: spec/controllers/sessions_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+        expect(response).to redirect_to(root_path)
+      end
+    end
+
+    describe '#create' do
+      before { get :create }
+
+      skip 'should create user if he does not exist in DB' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: same

File: spec/controllers/sessions_controller_spec.rb [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+      end
+    end
+
+    describe '#destroy' do
+      before do
+        expect(session[:user_id]).to eq(current_user.id)
+        delete :destroy
+      end
+
+      it 'should remove session' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: `it 'removes a user session'`

File: spec/controllers/sessions_controller_spec.rb [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+        expect(session[:user_id]).to be_nil
+      end
+
+      it 'should be redirected to log in page' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: same, no need to use `should` word if not needed

File: spec/models/user_spec.rb (hunk 17 above)
Comment (IlyaUmanets) [reply spec/models/user_spec.rb]: no need to spec `with_message` if this `message` is a default message

File: spec/models/user_spec.rb [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+    it { is_expected.to validate_presence_of(:provider).with_message("can't be blank") }
+    it { is_expected.to validate_presence_of(:image).with_message("can't be blank") }
+  end
+
+  describe '.find_or_create_user' do
+    it 'creates user if he dies not exist' do
+      expect { described_class.find_or_create_user(fields) }.to change(User, :count).by(1)
+    end
+
+    it 'returns user if he exists' do
```
Comment (IlyaUmanets) [inline spec/models/user_spec.rb]: `it 'returns user if exists'`

File: spec/controllers/sessions_controller_spec.rb [hunk 27]
```diff
@@ -0,0 +1,100 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+RSpec.describe SessionsController, type: :controller do
+  let(:current_user) { create(:user) }
+
+  describe '#auth_params' do
+    it 'returns only valid params' do
```
Comment (kolokolochek) [inline spec/controllers/sessions_controller_spec.rb]: How to do this spec?

File: spec/controllers/sessions_controller_spec.rb [hunk 28]
```diff
@@ -0,0 +1,100 @@
...
+      end
+    end
+
+    describe '#create' do
+      before { mock_auth_hash_response }
+
+      it 'creates user if does not exist' do
+        get :create, params: { provider: 'facebook' }
+
+        expect { '.....' }.to change(User, :count).by(1)
```
Comment (kolokolochek) [inline spec/controllers/sessions_controller_spec.rb]: How to do this spec?

File: spec/controllers/sessions_controller_spec.rb [hunk 29]
```diff
@@ -0,0 +1,100 @@
...
+        expect(response).to redirect_to(root_path)
+        expect(flash.notice).to eq('You have successfully logged in.')
+      end
+
+      it 'will find a user if already exist' do
+        expect(current_user).to be_persisted
+
+        get :create, params: { provider: 'facebook' }
+
+        expect { '.....' }.not_to change(User, :count)
```
Comment (kolokolochek) [inline spec/controllers/sessions_controller_spec.rb]: How to do this spec?

File: spec/controllers/sessions_controller_spec.rb (hunk 27 above)
Comment (IlyaUmanets) [reply spec/controllers/sessions_controller_spec.rb]: You should test `private` methods, test only `public` methods. `#create`, `#destroy`, `#failure`

File: spec/controllers/sessions_controller_spec.rb [hunk 30]
```diff
@@ -0,0 +1,100 @@
...
+  context 'when user is not authorized' do
+    describe '#log_in' do
+      it 'shows log_in page' do
+        get :log_in
+
+        expect(response).to render_template(:log_in)
+      end
+    end
+
+    describe '#create' do
```
Comment (IlyaUmanets) [inline spec/controllers/sessions_controller_spec.rb]: You need write 2 specs in this case. The first one when user is new and when user is persisted in the database. For the second you need to create user with parameters you specify in `mock_auth_hash_response` method

File: spec/controllers/sessions_controller_spec.rb (hunk 28 above)
Comment (IlyaUmanets) [reply spec/controllers/sessions_controller_spec.rb]: `expect { get :create, params: { provider: 'facebook' }.to change(User, :count).by(1)`

You can also expect response and notice in this `it`

**Pull request merged**